plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftLaunchController.java

```diff
@@ -173,93 +177,98 @@ private void sleep(long t) {
-						Connection conn = ConnectionsRegistryUtil.getConnectionFor(dc);
-						conn.updateResource(dc);
-					} catch (Exception e) {
-						String message = "Unable to enable 'DEV_MODE' for deployment config " + dc.getName(); //$NON-NLS-1$
-						OpenShiftCoreActivator.getDefault().getLogger().logError(message, e);
-						return new Status(Status.ERROR, OpenShiftCoreActivator.PLUGIN_ID, message, e);
-					}
+		if (!OpenShiftServerUtils.isNodeJsProject(server)) {
+			return;
+		}
+
+		new Job("Enabling " + DEV_MODE + " for deployment config " + dc.getName()) { //$NON-NLS-1$
+
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					dc.setEnvironmentVariable(DEV_MODE, Boolean.TRUE.toString());
+					Connection conn = ConnectionsRegistryUtil.getConnectionFor(dc);
+					conn.updateResource(dc);
 					return Status.OK_STATUS;
+				} catch (Exception e) {
+					String message = "Unable to enable " + DEV_MODE + " for deployment config " + dc.getName(); //$NON-NLS-1$
+					OpenShiftCoreActivator.getDefault().getLogger().logError(message, e);
+					return new Status(Status.ERROR, OpenShiftCoreActivator.PLUGIN_ID, message, e);
 				}
+			}
 
-			}.schedule();
-		}
+		}.schedule();
 	}
 
-	private void checkServerState(ControllableServerBehavior beh, String oldMode, String mode) {
+	private void setServerState(OpenShiftServerBehaviour beh, String oldMode, String mode) {
 		int state = pollState();
-		
+
 		if (!Objects.equals(oldMode, mode)) {
-			IModule[] modules = getServer().getModules();
-			for( int i = 0; i < modules.length; i++ ) {
-				((Server)getServer()).setModulePublishState(new IModule[]{modules[i]}, IServer.PUBLISH_STATE_FULL);
-			}
-			
-			if( state == IServer.STATE_STARTED && Boolean.TRUE.equals(beh.getSharedData(OpenShiftServerBehaviour.CURRENTLY_RESTARTING))) {
-				// Kick publish server job
-				Job j = new Job("Publishing server " + getServer().getName()) {
-					@Override
-					protected IStatus run(IProgressMonitor monitor) {
-						return getServer().publish(IServer.PUBLISH_INCREMENTAL, monitor);
-					}
-				};
-				j.schedule(3000);
-				beh.putSharedData(OpenShiftServerBehaviour.CURRENTLY_RESTARTING, null);
+			setModulesPublishing();
+			if (state == IServer.STATE_STARTED
+					&& beh.isRestarting()) {
+				publishServer();
+				beh.setRestarting(false);
 			}
 		}
 
-		
-		if( state == IServer.STATE_STARTED) {
+		if (state == IServer.STATE_STARTED) {
 			beh.setRunMode(mode);
 			beh.setServerStarted();
 		} else {
 			beh.setServerStopped();
-			((ControllableServerBehavior)getControllableBehavior()).setRunMode(null);
+			beh.setRunMode(null);
 		}
 	}
 
+	private void setModulesPublishing() {
+		IModule[] modules = getServer().getModules();
+		for (int i = 0; i < modules.length; i++) {
+			((Server) getServer()).setModulePublishState(
+					new IModule[] { modules[i] }, IServer.PUBLISH_STATE_FULL);
+		}
+	}
+	
+	private void publishServer() {
```

Comment:
I would comment this method by indicating it kicks off an incremental publish action in a job.